jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -2388,17 +2406,31 @@ public boolean equalsIgnoreCase(String string) {
 				// overloads rather than code points.
 				/*[ENDIF]*/
 				if (byteAtO1 != byteAtO2 
-						&& toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))
-						&& toLowerCase(helpers.byteToCharUnsigned(byteAtO1)) != toLowerCase(helpers.byteToCharUnsigned(byteAtO2))) {
+						&& toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))) {
 					return false;
 				}
 			}
 		} else {
-			while (o1 < end) {
+			// Compare the last chars.
+			// In order to tell 2 chars are different:
+			// If at least one char is ASCII, converting to upper cases then compare should be sufficient.
+			// If both chars are not in ASCII char set, need to convert to lower case and compare as well.
+			char charAtO1Last = s1.charAtInternal(s1len-1, s1Value);
+			char charAtO2Last = s2.charAtInternal(s1len-1, s2Value);
+
+			if (charAtO1Last != charAtO2Last 
+					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
+					&& (charAtO1Last < 256 || charAtO2Last < 256 || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))) {
```

Comment:
I think this test is wrong. Both characters need to be < 256.

Reply:
Great catch. You're right, the first upper-lower case mappings that're non-unique happens to step on ASCII character (i), at char (304) as you pointed out. So this check needs to be more strict to limit both char to ASCII, which guarantees 1-1 mapping.